// @tanstack/router-generator/dist/cjs/config.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const path = require("node:path");
const fs = require("node:fs");
const zod = require("zod");
const configSchema = zod.z.object({
  routeFilePrefix: zod.z.string().optional(),
  routeFileIgnorePrefix: zod.z.string().optional().default("-"),
  routeFileIgnorePattern: zod.z.string().optional(),
  routesDirectory: zod.z.string().optional().default("./src/routes"),
  generatedRouteTree: zod.z.string().optional().default("./src/routeTree.gen.ts"),
  quoteStyle: zod.z.enum(["single", "double"]).optional().default("single"),
  semicolons: zod.z.boolean().optional().default(false),
  disableTypes: zod.z.boolean().optional().default(false),
  addExtensions: zod.z.boolean().optional().default(false),
  disableLogging: zod.z.boolean().optional().default(false),
  disableManifestGeneration: zod.z.boolean().optional().default(false),
  apiBase: zod.z.string().optional().default("/api"),
  routeTreeFileHeader: zod.z.array(zod.z.string()).optional().default([
    "/* prettier-ignore-start */",
    "/* eslint-disable */",
    "// @ts-nocheck",
    "// noinspection JSUnusedGlobalSymbols"
  ]),
  routeTreeFileFooter: zod.z.array(zod.z.string()).optional().default(["/* prettier-ignore-end */"]),
  autoCodeSplitting: zod.z.boolean().optional(),
  indexToken: zod.z.string().optional().default("index"),
  routeToken: zod.z.string().optional().default("route"),
  experimental: zod.z.object({
    // TODO: Remove this option in the next major release (v2).
    enableCodeSplitting: zod.z.boolean().optional()
  }).optional()
});
function getConfig(inlineConfig = {}, configDirectory) {
  if (configDirectory === void 0) {
    configDirectory = process.cwd();
  }
  const configFilePathJson = path.resolve(configDirectory, "tsr.config.json");
  const exists = fs.existsSync(configFilePathJson);
  let config;
  if (exists) {
    config = configSchema.parse({
      ...JSON.parse(fs.readFileSync(configFilePathJson, "utf-8")),
      ...inlineConfig
    });
  } else {
    config = configSchema.parse(inlineConfig);
  }
  if (config.disableTypes) {
    config.generatedRouteTree = config.generatedRouteTree.replace(
      /\.(ts|tsx)$/,
      ".js"
    );
  }
  if (configDirectory) {
    if (path.isAbsolute(configDirectory)) {
      config.routesDirectory = path.resolve(
        configDirectory,
        config.routesDirectory
      );
      config.generatedRouteTree = path.resolve(
        configDirectory,
        config.generatedRouteTree
      );
    } else {
      config.routesDirectory = path.resolve(
        process.cwd(),
        configDirectory,
        config.routesDirectory
      );
      config.generatedRouteTree = path.resolve(
        process.cwd(),
        configDirectory,
        config.generatedRouteTree
      );
    }
  }
  validateConfig(config);
  return config;
}
function validateConfig(config) {
  var _a;
  if (typeof ((_a = config.experimental) == null ? void 0 : _a.enableCodeSplitting) !== "undefined") {
    const message = `
------
⚠️ ⚠️ ⚠️
ERROR: The "experimental.enableCodeSplitting" flag has been made stable and is now "autoCodeSplitting". Please update your configuration file to use "autoCodeSplitting" instead of "experimental.enableCodeSplitting".
------
`;
    console.error(message);
    throw new Error(message);
  }
  if (config.indexToken === config.routeToken) {
    throw new Error(
      `The "indexToken" and "routeToken" options must be different.`
    );
  }
  return config;
}
exports.configSchema = configSchema;
exports.getConfig = getConfig;
